[SCM-1022] - jgit push failure is not failing the build

attiand opened the pull request.

michael-o commented:
Looking into...

kwin commented on maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/test/java/org/apache/maven/scm/provider/git/jgit/command/checkin/JGitCheckinCommandTest.java:
```diff
@@ -0,0 +1,104 @@
+
+import static org.junit.Assert.assertFalse;
+
+public class JGitCheckinCommandTest extends ScmTestCase {
```
I would prefer having this test as part of the Git TCK so that it is executed against both Git CLI and JGit.

attiand replied:
Good point. It turns out that the `RemoteRefUpdate.Status#UP_TO_DATE` equivalent for gitexe is considered ok as the `git push` command returns zero. Which seems questionable for maven-release. Nothing would be changed in the upstream.

Which is better: To have equal behavior for both implementations, or a successful, possibly empty release?

kwin replied:
I think a push which doesn't add any commits should never be an error (neither in JGit nor in GitExe). If there is an up2date check being necessary in maven-release-plugin that needs to be triggered separately. I would therefore appreciate to add the test to the TCK but assert that it is returning a success status.

michael-o commented on maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/checkin/JGitCheckInCommand.java:
```diff
@@ -139,7 +141,20 @@ protected CheckInScmResult executeCheckInCommand(
+                            return new CheckInScmResult(
+                                    "JGit checkin",
+                                    "The git-push command failed, with status: "
+                                            + remoteRefUpdate.getStatus().name(),
```
I think we can reduce to `getStatus()` and the implicit string conv will do the magic for us.

michael-o commented on maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/test/java/org/apache/maven/scm/provider/git/jgit/command/checkin/JGitCheckinCommandTest.java:
```diff
@@ -0,0 +1,104 @@
+
+import static org.junit.Assert.assertFalse;
+
+public class JGitCheckinCommandTest extends ScmTestCase {
```
If this remains file name should be `JGitCheckInCommandTest`

michael-o commented:
This PR triggers https://github.com/apache/mina-sshd/issues/282. We first need to upgrade MINA. I am working on this right now.

attiand commented:
Moved RejectedNonFastForward Test to TCK. Removed UpToDate test since gitexec and jgit return different status, see https://github.com/apache/maven-scm/pull/203#discussion_r1551846585

michael-o reviewed (approved):
I am fine with this PR.

michael-o commented:
@kwin Any objections?

kwin commented on maven-scm-providers/maven-scm-providers-git/maven-scm-provider-jgit/src/main/java/org/apache/maven/scm/provider/git/jgit/command/checkin/JGitCheckInCommand.java:
```diff
@@ -139,7 +141,19 @@ protected CheckInScmResult executeCheckInCommand(
+
+                for (PushResult pushResult : pushResultList) {
+                    for (RemoteRefUpdate remoteRefUpdate : pushResult.getRemoteUpdates()) {
+                        if (remoteRefUpdate.getStatus() != RemoteRefUpdate.Status.OK) {
```
I think up2date is no error condition either: https://archive.eclipse.org/jgit/docs/jgit-2.0.0.201206130900-r/apidocs/org/eclipse/jgit/transport/RemoteRefUpdate.Status.html#UP_TO_DATE

attiand commented:
Status.UP_TO_DATE is now successful. Added TCK testUpToDatePush.

michael-o commented:
Let me run this again

jira-importer commented:
Resolve #1244
